### PR TITLE
Add SignalContext for graceful shutdown

### LINKED ISSUE #30: Implement server wide context for graceful shutdown

### PR DESCRIPTION
**Pull request by vlkolmakov**
Closes #30

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -23,6 +26,42 @@ func main() {
...
 
 	loggedMux := loghandler.NewLogHandler(mux)
-	log.Printf("server is listening at %s", port)
-	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", port), loggedMux))
+
+	log.Printf("server is listening at %s\n", port)
+
+	cfg.DebugContext = log.Printf
+
+	ctx := cfg.Context()
```
Comment (kepkap) [inline main.go:34]: i do not quite follow the need for custom context wrapper usage here. It adds extra complexity managing 'running' state, which i do not see the reason for.

Why we cannot just wait on channel? 

done := make(chan os.Signal, 1)
signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
<-done
call server.shutdown(here goes timeout context)

File: cfg/context.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
+// Copyright 2022 The Drivers Cooperative. All rights reserved.
+// Use of this source code is governed by a dual
+// license that can be found in the LICENSE file.
+
+package cfg
```
Comment (kepkap) [inline cfg/context.go:5]: package cfg implies containing something to do with config, in this regard cfg.context is misleading.

File: main.go (hunk 1 above)
Comment (vlkolmakov) [reply main.go:34]: Because proper interruptible context is not only used for awaiting for graceful shutdown. It is also used as a base context for every HTTP request (see http.Server struct). It is also used during request workflow to stop processing pipeline when context is interrupted/cancelled. It is also will be used for every Redis connection for the same purpose. It is also will be used with every Postgres connection.

In short, it is the only correct way to work with cancellation in Go.
 
The proposed implementation is not complex at all.

File: cfg/context.go (hunk 2 above)
Comment (vlkolmakov) [reply cfg/context.go:5]: "conf" package will be retired after lightweight configuration will be implemented. All new configuration logic will be put in "cfg" package. "conf" is too long and does not follow existing naming such as "api", "log".

Comment (kepkap) [reply cfg/context.go:5]: i meant maybe context had to be put under 'context' named package, not cfg, config, conf, etc. matter of taste i guess. not that much important.

File: main.go (hunk 1 above)
Comment (kepkap) [reply main.go:34]: In graceful shutdown server stops serving new requests, and completes all actively running requests. We should not cancel all outstanding requests. That's the thing - to wait until all of them served  - then die.

If instead we want to cancel all actively running requests we could use BaseContext with context.WithCancel. Standard contexts allow us to do that. 

@vlkolmakov I am not trying to

**Comment by kevinmichaelchen** [comment]
We should be calling [`GracefulStop()`](https://pkg.go.dev/google.golang.org/grpc#Server.GracefulStop) somewhere

File: cfg/context.go (hunk 2 above)
Comment (vlkolmakov) [reply cfg/context.go:5]: I though about this. Putting it in the "context" will slash with the std lib context package. Not a good idea. "cfg" is the most convenient place since it will be used almost everywhere either for config vars or context.

File: main.go (hunk 1 above)
Comment (vlkolmakov) [reply main.go:34]: @kepkap I don't understand what you're trying to say and how my approach contradicts the graceful shutdown workflow. Show me exact line number(s) where is my impl is incorrect.

contextWithCancel does not intercept OS signals.  Cancel func should be managed somewhere - this is what SignalContext is exactly doing.

And again, interruptible context should be an application level singleton, not only for HTTP server but for other users as well.

**Review by kepkap — APPROVED** [review]

File: main.go (hunk 1 above)
Comment (kepkap) [reply main.go:34]: ok cool, thanks for clarification.

**Review by kevinmichaelchen — APPROVED** [review]
